packages/network-controller/src/NetworkController.ts

```diff
@@ -66,32 +77,189 @@ export type NetworkMetadata = {
+   */
   chainId: Hex;
-  ticker: string;
-  nickname?: string;
-  rpcPrefs?: {
-    blockExplorerUrl: string;
-  };
+  /**
+   * A reference to a URL that the client will use by default to allow the user
+   * to view activity that has occurred on the chain. This index must refer to
+   * an item in `blockExplorerUrls`.
+   */
+  defaultBlockExplorerUrlIndex?: number;
+  /**
+   * A reference to an RPC endpoint that all requests will use by default in order to
+   * interact with the chain. This index must refer to an item in
+   * `rpcEndpoints`.
+   */
+  defaultRpcEndpointIndex: number;
+  /**
+   * The user-facing nickname assigned to the chain.
+   */
+  name: string;
+  /**
+   * The name of the currency to use for the chain.
+   */
+  nativeCurrency: string;
+  /**
+   * The collection of possible RPC endpoints that the client can use to
+   * interact with the chain.
+   */
+  rpcEndpoints: RpcEndpoint[];
+};
+
+/**
+ * A custom RPC endpoint in a new network configuration, meant to be used in
+ * conjunction with `AddNetworkFields`.
+ *
+ * Custom RPC endpoints do not need a `networkClientId` property because it is
+ * assumed that they have not already been added and therefore network clients
+ * do not exist for them yet (and hence IDs need to be generated).
+ */
+export type AddNetworkCustomRpcEndpointFields = Omit<
+  CustomRpcEndpoint,
+  'networkClientId'
+>;
+
+/**
+ * A new network configuration that `addNetwork` takes.
+ *
+ * Custom RPC endpoints do not need a `networkClientId` property because it is
+ * assumed that they have not already been added and are not represented by
+ * network clients yet.
+ */
+export type AddNetworkFields = Omit<NetworkConfiguration, 'rpcEndpoints'> & {
+  rpcEndpoints: (InfuraRpcEndpoint | AddNetworkCustomRpcEndpointFields)[];
 };
 
 /**
- * The collection of network configurations in state.
+ * A custom RPC endpoint in an updated representation of a network
+ * configuration, meant to be used in conjunction with `UpdateNetworkFields`.
+ *
+ * Custom RPC endpoints do not need a `networkClientId` property because it is
+ * assumed that they have not already been added and therefore network clients
+ * do not exist for them yet (and hence IDs need to be generated).
  */
-type NetworkConfigurations = Record<
-  NetworkConfigurationId,
-  NetworkConfiguration & { id: NetworkConfigurationId }
+export type UpdateNetworkCustomRpcEndpointFields = Partialize<
+  CustomRpcEndpoint,
+  'networkClientId'
 >;
 
+/**
+ * An updated representation of an existing network configuration that
+ * `updateNetwork` takes.
+ *
+ * Custom RPC endpoints may or may not have a `networkClientId` property; if
+ * they do, then it is assumed that they already exist, and if not, then it is
+ * assumed that they are new and are not represented by network clients yet.
+ */
+type UpdateNetworkFields = Omit<NetworkConfiguration, 'rpcEndpoints'> & {
+  rpcEndpoints: (InfuraRpcEndpoint | UpdateNetworkCustomRpcEndpointFields)[];
```

Comment:
does this mean that only rpcEndpoints is updatable ?

Reply:
No, all properties are updatable. This type mixes in the same properties that `NetworkConfiguration` has, but overrides `rpcEndpoints` so that `networkClientId` is an optional property for each RPC endpoint instead of being required.